File: visualstudio-extension/src/Atec.SourceTrace.Core/TraceRequestBuilder.cs
using System;
using System.Collections.Generic;

namespace Atec.SourceTrace.Core;

/// <summary>Builds request JSON for Backend v2.6 — mirrors vscode-extension requestBuilder.ts.</summary>
public static class TraceRequestBuilder
{
    public const int DefaultMaxSelectedChars = 4000;
    public const string DefaultQuery = "선택한 코드가 왜 변경됐는지 알려줘";

    public sealed class TruncateResult
    {
        public string Code { get; }
        public bool Truncated { get; }

        public TruncateResult(string code, bool truncated)
        {
            Code = code;
            Truncated = truncated;
        }
    }

    public sealed class BuiltRequest
    {
        public string JsonBody { get; }
        public bool Truncated { get; }

        public BuiltRequest(string jsonBody, bool truncated)
        {
            JsonBody = jsonBody;
            Truncated = truncated;
        }
    }

    public static TruncateResult TruncateSelectedCode(string? code, int maxChars)
    {
        if (code == null)
        {
            return new TruncateResult(string.Empty, false);
        }

        if (code.Length <= maxChars)
        {
            return new TruncateResult(code, false);
        }

        return new TruncateResult(code.Substring(0, maxChars), true);
    }

    public static BuiltRequest BuildReportRequest(
        int equipmentId,
        string? query,
        string filePath,
        string? selectedCode,
        string? detectedSymbol,
        string? sourceMode,
        bool useOllama,
        int maxSelectedCodeChars)
    {
        var selectedText = (selectedCode ?? string.Empty).Trim();
        var symbol = string.IsNullOrWhiteSpace(detectedSymbol) ? null : detectedSymbol.Trim();
        var selectedForBody = selectedText.Length > 0 ? selectedText : (symbol ?? string.Empty);
        var augmentedQuery = SymbolExtractor.AugmentQueryWithSymbol(
            string.IsNullOrWhiteSpace(query) ? DefaultQuery : query.Trim(),
            symbol);
        var sent = TruncateSelectedCode(selectedForBody, maxSelectedCodeChars);

        var body = new Dictionary<string, object?>
        {
            ["equipment_id"] = equipmentId,
            ["query"] = augmentedQuery,
            ["file_path"] = filePath,
            ["selected_code"] = sent.Code,
            ["use_ollama"] = useOllama,
        };

        if (!string.IsNullOrEmpty(sourceMode))
        {
            body["source_mode"] = sourceMode;
        }

        if (symbol != null)
        {
            body["detected_symbol"] = symbol;
        }

        var json = JsonUtil.Object(body);
        if (string.IsNullOrWhiteSpace(json) || json == "{}")
        {
            throw new InvalidOperationException("report request body가 비어 있습니다.");
        }

        if (!json.Contains("\"equipment_id\""))
        {
            throw new InvalidOperationException("report request에 equipment_id가 없습니다.");
        }

        return new BuiltRequest(json, sent.Truncated);
    }

    public static BuiltRequest BuildSelectionRequest(
        int equipmentId,
        string? repoRelativePath,
        int? repoIdHint,
        string? clientFilePath,
        int startLine,
        int endLine,
        string? selectedCode,
        string? enclosingSymbol,
        int maxSelectedCodeChars,
        string? revision)
    {
        var sent = TruncateSelectedCode(selectedCode ?? string.Empty, maxSelectedCodeChars);
        var body = new Dictionary<string, object?>
        {
            ["equipment_id"] = equipmentId,
            ["start_line"] = startLine,
            ["end_line"] = endLine,
            ["selected_code"] = sent.Code,
            ["revision"] = string.IsNullOrWhiteSpace(revision) ? "HEAD" : revision.Trim(),
        };

        if (!string.IsNullOrWhiteSpace(enclosingSymbol))
        {
            body["enclosing_symbol"] = enclosingSymbol.Trim();
        }

        var rel = (repoRelativePath ?? string.Empty).Trim().Replace('\\', '/');
        if (rel.Length > 0)
        {
            body["repo_relative_path"] = rel;
            if (repoIdHint.HasValue && repoIdHint.Value > 0)
            {
                body["repo_id"] = repoIdHint.Value;
                body["repo_id_hint"] = repoIdHint.Value;
            }

            if (!string.IsNullOrWhiteSpace(clientFilePath))
            {
                body["client_file_path"] = clientFilePath;
            }
        }
        else if (!string.IsNullOrWhiteSpace(clientFilePath))
        {
            body["file_path"] = clientFilePath;
        }

        return new BuiltRequest(JsonUtil.Object(body), sent.Truncated);
    }

    public static string PickResultMarkdown(string? responseJson)
    {
        if (string.IsNullOrWhiteSpace(responseJson))
        {
            return "(빈 응답)";
        }

        foreach (var key in new[] { "content", "answer", "evidence_answer" })
        {
            var value = JsonUtil.GetStringField(responseJson, key);
            if (!string.IsNullOrWhiteSpace(value))
            {
                return value;
            }
        }

        return "```json\n" + responseJson + "\n```";
    }

    public static bool LooksLikeAmbiguity(string? messageOrBody)
    {
        if (string.IsNullOrEmpty(messageOrBody))
        {
            return false;
        }

        return messageOrBody.Contains("여러 장비 Repository")
               || messageOrBody.IndexOf("AMBIGUOUS", StringComparison.OrdinalIgnoreCase) >= 0
               || messageOrBody.Contains("하나를 결정할 수 없")
               || messageOrBody.Contains("동일한 파일 경로가 여러");
    }
}

File: eclipse-plugin/com.atec.sourcetrace.eclipse/src/com/atec/sourcetrace/eclipse/core/TraceRequestBuilder.java
package com.atec.sourcetrace.eclipse.core;

import java.util.Map;

/**
 * Builds request JSON for Backend v2.6 contracts — mirrors
 * vscode-extension {@code requestBuilder.ts}.
 */
public final class TraceRequestBuilder {

	public static final int DEFAULT_MAX_SELECTED_CHARS = 4000;
	public static final String DEFAULT_QUERY = "선택한 코드가 왜 변경됐는지 알려줘";

	public static final class TruncateResult {
		public final String code;
		public final boolean truncated;

		public TruncateResult(String code, boolean truncated) {
			this.code = code;
			this.truncated = truncated;
		}
	}

	public static final class BuiltRequest {
		public final String jsonBody;
		public final boolean truncated;

		public BuiltRequest(String jsonBody, boolean truncated) {
			this.jsonBody = jsonBody;
			this.truncated = truncated;
		}
	}

	private TraceRequestBuilder() {
	}

	public static TruncateResult truncateSelectedCode(String code, int maxChars) {
		if (code == null) {
			return new TruncateResult("", false);
		}
		if (code.length() <= maxChars) {
			return new TruncateResult(code, false);
		}
		return new TruncateResult(code.substring(0, maxChars), true);
	}

	public static BuiltRequest buildReportRequest(
			int equipmentId,
			String query,
			String filePath,
			String selectedCode,
			String detectedSymbol,
			String sourceMode,
			boolean useOllama,
			int maxSelectedCodeChars) {
		String selectedText = selectedCode == null ? "" : selectedCode.trim();
		String symbol = detectedSymbol == null ? null : detectedSymbol.trim();
		if (symbol != null && symbol.isEmpty()) {
			symbol = null;
		}
		String selectedForBody = !selectedText.isEmpty() ? selectedText
				: (symbol != null ? symbol : "");
		String augmentedQuery = SymbolExtractor.augmentQueryWithSymbol(
				query == null || query.trim().isEmpty() ? DEFAULT_QUERY : query.trim(),
				symbol);
		TruncateResult sent = truncateSelectedCode(selectedForBody, maxSelectedCodeChars);

		Map<String, Object> body = JsonUtil.mapOf();
		body.put("equipment_id", Integer.valueOf(equipmentId));
		body.put("query", augmentedQuery);
		body.put("file_path", filePath);
		body.put("selected_code", sent.code);
		body.put("use_ollama", Boolean.valueOf(useOllama));
		if (sourceMode != null && !sourceMode.isEmpty()) {
			body.put("source_mode", sourceMode);
		}
		if (symbol != null) {
			body.put("detected_symbol", symbol);
		}
		String json = JsonUtil.object(body);
		if (json == null || json.isBlank() || "{}".equals(json.trim())) {
			throw new IllegalStateException("report request body가 비어 있습니다.");
		}
		if (!json.contains("\"equipment_id\"")) {
			throw new IllegalStateException("report request에 equipment_id가 없습니다.");
		}
		return new BuiltRequest(json, sent.truncated);
	}

	public static BuiltRequest buildSelectionRequest(
			int equipmentId,
			String repoRelativePath,
			Integer repoIdHint,
			String clientFilePath,
			int startLine,
			int endLine,
			String selectedCode,
			String enclosingSymbol,
			int maxSelectedCodeChars,
			String revision) {
		TruncateResult sent = truncateSelectedCode(
				selectedCode == null ? "" : selectedCode,
				maxSelectedCodeChars);
		Map<String, Object> body = JsonUtil.mapOf();
		body.put("equipment_id", Integer.valueOf(equipmentId));
		body.put("start_line", Integer.valueOf(startLine));
		body.put("end_line", Integer.valueOf(endLine));
		body.put("selected_code", sent.code);
		body.put("revision", revision == null || revision.trim().isEmpty() ? "HEAD" : revision.trim());
		if (enclosingSymbol != null && !enclosingSymbol.trim().isEmpty()) {
			body.put("enclosing_symbol", enclosingSymbol.trim());
		}
		String rel = repoRelativePath == null ? "" : repoRelativePath.trim().replace('\\', '/');
		if (!rel.isEmpty()) {
			body.put("repo_relative_path", rel);
			if (repoIdHint != null && repoIdHint.intValue() > 0) {
				body.put("repo_id", repoIdHint);
				body.put("repo_id_hint", repoIdHint);
			}
			if (clientFilePath != null && !clientFilePath.trim().isEmpty()) {
				body.put("client_file_path", clientFilePath);
			}
		} else if (clientFilePath != null && !clientFilePath.trim().isEmpty()) {
			body.put("file_path", clientFilePath);
		}
		return new BuiltRequest(JsonUtil.object(body), sent.truncated);
	}

	/** Prefer Backend Markdown fields in the same order as VS Code pickResultMarkdown. */
	public static String pickResultMarkdown(String responseJson) {
		if (responseJson == null || responseJson.trim().isEmpty()) {
			return "(빈 응답)";
		}
		for (String key : new String[] { "content", "answer", "evidence_answer" }) {
			String value = JsonUtil.getStringField(responseJson, key);
			if (value != null && !value.trim().isEmpty()) {
				return value;
			}
		}
		return "```json\n" + responseJson + "\n```";
	}

	public static boolean looksLikeAmbiguity(String messageOrBody) {
		if (messageOrBody == null) {
			return false;
		}
		String m = messageOrBody;
		return m.contains("여러 장비 Repository")
				|| m.contains("AMBIGUOUS")
				|| m.contains("하나를 결정할 수 없")
				|| m.contains("동일한 파일 경로가 여러");
	}
}
